rest-client/client-vertx-impl/src/main/java/io/apicurio/registry/rest/client/VertxHttpClient.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright 2021 Red Hat
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.apicurio.registry.rest.client;
+
+import io.apicurio.registry.auth.Auth;
+import io.apicurio.registry.rest.client.config.ClientConfig;
+import io.apicurio.registry.rest.client.exception.RestClientException;
+import io.apicurio.registry.rest.client.impl.ErrorHandler;
+import io.apicurio.registry.rest.client.request.Request;
+import io.apicurio.registry.rest.client.response.ResponseHandler;
+import io.apicurio.registry.utils.IoUtil;
+import io.vertx.core.Vertx;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.http.HttpMethod;
+import io.vertx.core.http.RequestOptions;
+import io.vertx.ext.web.client.HttpRequest;
+import io.vertx.ext.web.client.WebClient;
+import org.apache.http.NameValuePair;
+import org.apache.http.client.utils.URIBuilder;
+import org.keycloak.authorization.client.util.HttpResponseException;
+
+import java.io.UncheckedIOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+
+public class VertxHttpClient implements RegistryHttpClient {
+
+    private final WebClient webClient;
+    private final Auth auth;
+    private final String basePath;
+    private final Vertx vertx;
+
+    private static final Map<String, String> DEFAULT_HEADERS = new HashMap<>();
+    private static final ThreadLocal<Map<String, String>> requestHeaders = ThreadLocal.withInitial(Collections::emptyMap);
+
+    public VertxHttpClient(String basePath, Map<String, Object> options, Auth auth) {
+        if (!basePath.endsWith("/")) {
+            basePath += "/";
+        }
+        this.vertx = Vertx.currentContext() != null ? Vertx.currentContext().owner() : Vertx.vertx();
```

Comment:
I'm no Vert.x expert, but this sounds wrong, because you're creating a completely new Vert.x instance (the `Vertx.vertx()` call) if you're not on a Vert.x thread. I'd suggest the constructor should take `Vertx` as an argument.

Reply:
I'm creating a vertx instance just if no context can be found. Due to the SPI implementation, it is hard to add the vertex argument to the constructor without polluting the interfaces. That said, another option is to add a new constructor for external usage, does that work for you? I'm also interested to know there's a proper way to get the current instance other than the one implemented here.

Reply:
Ah I forgot about the factory thing. I'd say that being able to inject an implementation from the outside, instead of looking it up yourself via `ServiceLoader` would work better for Quarkus.

The way you lookup the Vert.x instance for current thread is OK, but here's a catch: there's no global Vert.x instance. You can have multiple Vert.x instances running at the same time, each having different set of thread pools. So depending on from which thread your constructor is called, there may be an associated Vert.x context, or not.

Reply:
Ok, I will add both a new constructor to the Vert.x client and a new method to the factory to be able to do both, inject a httpclient and sending the Vert.x instance.